File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        
        int l=1;
        int r=piles[0];
        for(int i:piles){
            r=Math.max(r,i);
        }
        while(l<r){
            int m=(l+r)/2;
            if(canFinish(h,piles,m)){
                r=m;
            }else{
                l=m+1;
            }
        }
        return l;
    }
    boolean canFinish(int h,int []piles,int m){
        int c=0;
        for(int i:piles){
            c+=i/m;
            if(i%m!=0){
                c++;
            }
        }
        return c<=h;
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l=1
        r=0
        for i in range (len(piles)):
            r=max(r,piles[i])

        while(l<r):
            m=l+(r-l)//2
            if(self.isPos(m,piles,h)):
                r=m
            else:
                l=m+1
        return l
    def isPos(self,m,piles,h):
        c=0
        for i in piles:
            c=c+i//m
            if i%m!=0:
                c=c+1
        return c<=h
